proof/refine/ARM/Ipc_R.thy

```diff
@@ -2990,6 +2990,163 @@ lemma do_nbrecv_failed_transfer_corres:
+  "thread_set f tptr \<lbrace>valid_tcb ptr tcb\<rbrace>"
+  apply (clarsimp simp: thread_set_def)
+  apply_trace (wpsimp wp: get_object_wp)
+  done
+
+lemma thread_set_valid_tcbs[wp]:
+ "\<lbrace>valid_tcbs and tcb_at tptr and
+    (\<lambda>s. \<forall>tcb. valid_tcb tptr tcb s \<longrightarrow> valid_tcb tptr (f tcb) s)\<rbrace>
+   thread_set f tptr
+   \<lbrace>\<lambda>_. valid_tcbs\<rbrace>"
+  unfolding valid_tcbs_def
+  apply (wpsimp wp: hoare_vcg_all_lift hoare_vcg_imp_lift')
+  apply (clarsimp simp: obj_at_def is_tcb)
+  done
+
+lemma valid_tcbs_valid_tcbE:
+  assumes "tcb_at t s"
+          "valid_tcbs s"
+          "\<And>tcb. ko_at (TCB tcb) t s \<Longrightarrow> valid_tcb t tcb s \<Longrightarrow> R s (TCB tcb)"
+  shows "obj_at (R s) t s"
+  using assms
+  apply (clarsimp simp: obj_at_def)
+  apply (rename_tac ko)
+  apply (case_tac ko; clarsimp simp: is_tcb_def)
+  apply (rename_tac tcb)
+  apply (prop_tac "valid_tcb t tcb s")
+   apply (clarsimp simp: valid_tcbs_def)
+   apply (drule_tac x=t in spec)
+   apply (drule_tac x=tcb in spec)
+   apply (clarsimp simp: obj_at_def)
+   apply (clarsimp simp: )
+  done
+
+lemma cross_rel_srE:
+  "\<lbrakk> (s, s') \<in> sr; cross_rel_ul sr P Q; P s \<longrightarrow> Q s' \<Longrightarrow> R \<rbrakk> \<Longrightarrow> R"
+  by (simp add: cross_rel_ul_def)
+
+lemma thread_set_weak_valid_sched_action2:
+  "\<lbrace>weak_valid_sched_action and scheduler_act_not tptr\<rbrace> thread_set f tptr \<lbrace>\<lambda>rv. weak_valid_sched_action\<rbrace>"
+  apply (wpsimp wp: thread_set_wp simp: obj_at_kh_kheap_simps vs_all_heap_simps fun_upd_def
+                    weak_valid_sched_action_def)
+  apply (auto simp: weak_valid_sched_action_def scheduler_act_not_def)
+  apply (rule_tac x=ref' in exI; simp)
+  apply clarsimp
+ done
+
+definition notQueued :: "obj_ref \<Rightarrow> kernel_state \<Rightarrow> bool" where
+  "notQueued t s \<equiv> (\<forall>a b. t \<notin> set (ksReadyQueues s (a, b)))"
+
+lemma notQueued_cross_rel:
+  "cross_rel (not_queued t) (notQueued t)"
+  unfolding cross_rel_def state_relation_def
+  by (clarsimp simp: notQueued_def ready_queues_relation_def not_queued_def)
+
+lemma valid_tcb_sched_context_update_empty[elim!]:
+  "valid_tcb tp tcb s \<Longrightarrow> valid_tcb tp (tcb_sched_context_update Map.empty tcb) s"
+  by (auto simp: valid_tcb_def tcb_cap_cases_def)
+
+lemma valid_tcb'_SchedContext_update_empty[elim!]:
+  "valid_tcb' tcb s' \<Longrightarrow> valid_tcb' (tcbSchedContext_update Map.empty tcb) s'"
+  by (auto simp: valid_tcb'_def valid_cap'_def tcb_cte_cases_def)
+
+lemma maybeReturnSc_corres:
+  "corres dc
+   (ntfn_at ntfnPtr and tcb_at thread and valid_tcbs and pspace_aligned
+      and scheduler_act_not thread
+      and pspace_distinct and weak_valid_sched_action
+      and not_queued thread and not_in_release_q thread)
+   (valid_tcbs' and valid_queues and valid_queues' and valid_release_queue and valid_release_queue')
```

Comment:
maybe use `valid_release_queue_iff`? We probably could do with the same for `valid_queues` and the other...